drivers/usb/uhc/uhc_mcux_ehci.c

```diff
@@ -0,0 +1,427 @@
+	    (priv->controller_id <= kUSB_ControllerEhci1)) {
+		thread_entry = uhc_mcux_thread;
+	}
+
+	if (thread_entry == NULL) {
+		return -ENOMEM;
+	}
+
+#ifdef CONFIG_DT_HAS_NXP_USBPHY_ENABLED
+	if (config->phy_config != NULL) {
+		USB_EhciPhyInit(priv->controller_id, 0u, config->phy_config);
+	}
+#endif
+
+	priv->dev = dev;
+	/* Init MCUX USB host driver. */
+	priv->mcux_host.deviceCallback = mcux_host_callback;
+	status = priv->mcux_if->controllerCreate(priv->controller_id, &priv->mcux_host,
+						 &(priv->mcux_host.controllerHandle));
+	if (status != kStatus_USB_Success) {
+		return -ENOMEM;
+	}
+
+	/* Create MCUX USB host driver task */
+	k_thread_create(&priv->drv_stack_data, config->drv_stack, CONFIG_UHC_NXP_THREAD_STACK_SIZE,
+			thread_entry, (void *)dev, NULL, NULL, K_PRIO_COOP(2), 0, K_NO_WAIT);
+	k_thread_name_set(&priv->drv_stack_data, "uhc_mcux_ehci");
+
+	return 0;
+}
+
+static int uhc_mcux_enable(const struct device *dev)
+{
+	const struct uhc_mcux_config *config = dev->config;
+
+	config->irq_enable_func(dev);
+
+	return 0;
+}
+
+static int uhc_mcux_disable(const struct device *dev)
+{
+	const struct uhc_mcux_config *config = dev->config;
+
+	config->irq_disable_func(dev);
+
+	return 0;
+}
+
+static int uhc_mcux_shutdown(const struct device *dev)
+{
+	const struct uhc_mcux_config *config = dev->config;
+	struct uhc_mcux_data *priv = uhc_get_private(dev);
+
+	config->irq_disable_func(dev);
+	k_thread_abort(&priv->drv_stack_data);
+	priv->mcux_if->controllerDestory(priv->mcux_host.controllerHandle);
+
+	return 0;
+}
+
+/* Signal bus reset, 50ms SE0 signal */
+static int uhc_mcux_bus_reset(const struct device *dev)
+{
+	int ret;
+
+	ret = uhc_mcux_bus_control(dev, kUSB_HostBusReset);
+	if (ret == 0) {
+		ret = uhc_mcux_bus_control(dev, kUSB_HostBusRestart);
+	}
+
+	return ret;
+}
+
+/* Enable SOF generator */
+static int uhc_mcux_sof_enable(const struct device *dev)
+{
+	/* MCUX doesn't need it. */
+	return 0;
+}
+
+/* Disable SOF generator and suspend bus */
+static int uhc_mcux_bus_suspend(const struct device *dev)
+{
+	return uhc_mcux_bus_control(dev, kUSB_HostBusSuspend);
+}
+
+/* Signal bus resume event, 20ms K-state + low-speed EOP */
+static int uhc_mcux_bus_resume(const struct device *dev)
+{
+	return uhc_mcux_bus_control(dev, kUSB_HostBusResume);
+}
+
+/* MCUX hal controller driver transfer callback */
+static void uhc_mcux_transfer_callback(void *param, usb_host_transfer_t *transfer,
+				       usb_status_t status)
+{
+	const struct device *dev = (const struct device *)param;
+	struct uhc_transfer *const xfer = (struct uhc_transfer *const)(transfer->uhc_xfer);
+	int err = -EIO;
+
+	if (status == kStatus_USB_Success) {
+		err = 0;
+
+		if ((xfer->ep == 0) && (transfer->setupPacket->bRequest == USB_SREQ_SET_ADDRESS)) {
+			usb_host_pipe_handle mcux_ep_handle;
+
+			/* set address to let USB_HostHelperGetPeripheralInformation get right value
+			 */
+			xfer->addr = (uint8_t)transfer->setupPacket->wValue;
```

Comment:
xfer->udev->addr, but driver must not write to xfer->udev->addr.

Reply:
Updated.